### PR TITLE
rebase

### PR DESCRIPTION
**Pull request by levnach**

### CODE REVIEW

File: src/ast/rewriter/arith_rewriter.cpp [hunk 1]
```diff
@@ -800,107 +800,9 @@ br_status arith_rewriter::mk_idiv_core(expr * arg1, expr * arg2, expr_ref & resu
         result = m_util.mk_numeral(div(v1, v2), is_int);
         return BR_DONE;
     }
-    if (m_util.is_numeral(arg2, v2, is_int) && v2.is_one()) {
```
Comment (NikolajBjorner) [inline src/ast/rewriter/arith_rewriter.cpp]: this is new code and should be added into your rebase.

File: src/ast/rewriter/arith_rewriter.cpp [hunk 2]
```diff
@@ -800,107 +800,9 @@ br_status arith_rewriter::mk_idiv_core(expr * arg1, expr * arg2, expr_ref & resu
...
-        return BR_REWRITE_FULL;
-    }
     return BR_FAILED;
 }
 
-
-// 
-// implement div ab ac = floor( ab / ac) = floor (b / c) = div b c
-//
-bool arith_rewriter::divides(expr* num, expr* den, expr_ref& result) {
```
Comment (NikolajBjorner) [inline src/ast/rewriter/arith_rewriter.cpp]: ditto, add this to rebase

File: src/ast/rewriter/arith_rewriter.h [hunk 3]
```diff
@@ -95,10 +95,6 @@ class arith_rewriter : public poly_rewriter<arith_rewriter_core> {
     expr_ref neg_monomial(expr * e) const;
     expr * mk_sin_value(rational const & k);
     app * mk_sqrt(rational const & k);
-    bool divides(expr* d, expr* n, expr_ref& result);
```
Comment (NikolajBjorner) [inline src/ast/rewriter/arith_rewriter.h]: needs to be added to rebase

File: src/ast/simplifier/arith_simplifier_plugin.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
+/*++
```
Comment (NikolajBjorner) [inline src/ast/simplifier/arith_simplifier_plugin.h]: remove all files from "simplifier" directory

File: src/shell/main.cpp [hunk 5]
```diff
@@ -296,6 +296,9 @@ void parse_cmd_line_args(int argc, char ** argv) {
 
 int STD_CALL main(int argc, char ** argv) {
     try{
+        DEBUG_CODE( for (int i = 0; i < argc; i++)
```
Comment (NikolajBjorner) [inline src/shell/main.cpp]: remove

File: src/smt/smt_setup.cpp [hunk 6]
```diff
@@ -968,7 +979,7 @@ namespace smt {
         if (st.num_theories() == 2 && st.has_uf() && is_arith(st)) {
             if (!st.m_has_real)
                 setup_QF_UFLIA(st);
-            else if (!st.m_has_int && st.m_num_non_linear == 0)
```
Comment (NikolajBjorner) [inline src/smt/smt_setup.cpp]: add this to rebase

File: src/ast/rewriter/arith_rewriter.cpp (hunk 2 above)
Comment (levnach) [reply src/ast/rewriter/arith_rewriter.cpp]: If I add this code it would not compile.

File: src/util/lp/binary_heap_priority_queue_def.h [hunk 7]
```diff
@@ -20,7 +20,7 @@ Revision History:
 #include "util/vector.h"
 #include "util/lp/binary_heap_priority_queue.h"
 namespace lp {
-// this is the child place in the heap
+// is is the child place in heap
```
Comment (NikolajBjorner) [inline src/util/lp/binary_heap_priority_queue_def.h]: grammar?

File: src/util/lp/binary_heap_upair_queue_def.h [hunk 8]
```diff
@@ -29,7 +29,7 @@ template <typename T> binary_heap_upair_queue<T>::binary_heap_upair_queue(unsign
 
 template <typename T> unsigned
 binary_heap_upair_queue<T>::dequeue_available_spot() {
-    SASSERT(m_available_spots.empty() == false);
+    lp_assert(m_available_spots.empty() == false);
```
Comment (NikolajBjorner) [inline src/util/lp/binary_heap_upair_queue_def.h]: aka lp_assert(!m_available_spots.empty());

File: src/util/lp/bound_propagator.cpp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    if (strict)
+        kind = static_cast<lconstraint_kind>(kind / 2);
+    
+    if (!bound_is_interesting(j, kind, v))
+        return;
+     unsigned k; // index to ibounds
+     if (is_low) {
+         if (try_get_value(m_improved_lower_bounds, j, k)) {
+             auto & found_bound = m_ibounds[k];
+             if (v > found_bound.m_bound || (v == found_bound.m_bound && found_bound.m_strict == false && strict)) {
```
Comment (NikolajBjorner) [inline src/util/lp/bound_propagator.cpp]: aka !found_bound.m_strict

File: src/util/lp/core_solver_pretty_printer_def.h [hunk 10]
```diff
@@ -38,7 +38,7 @@ core_solver_pretty_printer<T, X>::core_solver_pretty_printer(lp_core_solver_base
     m_rs(ncols(), zero_of_type<X>()),
     m_w_buff(core_solver.m_w),
     m_ed_buff(core_solver.m_ed) {
-    m_low_bounds_title = "low";
+    m_lower_bounds_title = "low";
     m_upp_bounds_title = "upp";
```
Comment (NikolajBjorner) [inline src/util/lp/core_solver_pretty_printer_def.h]: low was changed to lower, but upp was not changed.

File: src/util/mpz.cpp [hunk 11]
```diff
@@ -146,11 +144,6 @@ mpz_manager<SYNCH>::mpz_manager():
     else {
         m_init_cell_capacity = 6;
     }
-    for (unsigned i = 0; i < 2; i++) {
```
Comment (NikolajBjorner) [inline src/util/mpz.cpp]: Seems that we will need to copy over the initialization of m_tmp[I], Only m_args was removed.

for (unsigned i = 0; i < 2; i++) { 
   m_tmp[i] = allocate(m_init_cell_capacity); 
}

File: src/util/lp/binary_heap_upair_queue_def.h (hunk 8 above)
Comment (levnach) [reply src/util/lp/binary_heap_upair_queue_def.h]: I find my version a bit more readable.

File: src/util/lp/core_solver_pretty_printer_def.h (hunk 10 above)
Comment (levnach) [reply src/util/lp/core_solver_pretty_printer_def.h]: The short names snug nicely into a column

File: src/util/mpz.cpp (hunk 11 above)
Comment (levnach) [reply src/util/mpz.cpp]: If this place is OK now?

Comment (NikolajBjorner) [reply src/util/mpz.cpp]: my comment was misguided.

File: src/smt/theory_lra.cpp [hunk 12]
```diff
@@ -2612,12 +2612,26 @@ class theory_lra::imp {
     }    
 
     bool get_lower(enode* n, expr_ref& r) {
-        NOT_IMPLEMENTED_YET();
+        theory_var v = n->get_th_var(get_id());
```
Comment (NikolajBjorner) [inline src/smt/theory_lra.cpp]: if (!can_get_value(v))) return false;
lp::var_index vi = m_theory_var2var_index[v];

…

if (m_solver->has_lower_bound(vi, ci, val, is_strict))
...

**Comment by NikolajBjorner** [comment]
superseded by other pull request.

**Pull request closed without merging**